mesh/submesh/ncsubmesh.hpp

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) 2010-2024, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_NCSUBMESH
+#define MFEM_NCSUBMESH
+
+#include "../ncmesh.hpp"
+#include "submesh.hpp"
+#include "submesh_utils.hpp"
+#include <unordered_map>
+
+namespace mfem
+{
+
+/**
+ * @brief Class representing a Nonconformal SubMesh. This is only used by SubMesh.
+ */
+class NCSubMesh : public NCMesh
+{
+   friend class SubMesh; ///< Only SubMesh can use methods in this class
+public:
+   using From = SubMesh::From;
+   /**
+    * @brief Get the parent Mesh object
+    *
+    */
+   const NCMesh* GetParent() const
+   {
+      return parent_;
+   }
+
+   /**
+    * @brief Get the From indicator.
+    *
+    * Indicates whether the SubMesh has been created from a domain or
+    * surface.
+    */
+   SubMesh::From GetFrom() const
+   {
+      return from_;
+   }
+
+   /**
+   * @brief Check if Mesh @a m is a SubMesh.
+   *
+   * @param m The input Mesh
+   */
+   static bool IsNCSubMesh(const NCMesh *m)
+   {
+      return dynamic_cast<const NCSubMesh *>(m) != nullptr;
+   }
+private:
+
+   /// Private constructor
+   NCSubMesh(SubMesh& submesh, const NCMesh &parent, From from,
+             const Array<int> &attributes);
+
+   /// The parent Mesh. Not owned.
+   const NCMesh *parent_;
+
+   /// Indicator from which part of the parent ParMesh the ParSubMesh is going
+   /// to be created.
+   From from_;
+
+   /// Attributes on the parent NCMesh on which the NCSubMesh is created.
+   /// Could either be domain or boundary attributes (determined by from_).
+   Array<int> attributes_;
+
+   /// Mapping from submesh element nc ids (index of the array), to the parent
+   /// element ids. If from a boundary, these map to faces in the parent.
+   Array<int> parent_element_ids_;
```

Comment:
The `NCMesh::Elements` are different to those of the mesh, so the map here has to be different to the one in the `SubMesh`, and there has to be duplication across `NCSubMesh` and `ParNCSubMesh` because of the weird inheritance/composition structure of Mesh <-> NCMesh. 

I removed the extra unneeded member variables in https://github.com/mfem/mfem/pull/4476/commits/9f708051daefaacde29fc2ba0447050eeb625fa2